docs/celestia-architecture/adr-010-reactor-specific-peers.md

```diff
@@ -0,0 +1,173 @@
+
+#### Key Features:
+- Reactors can define customized criteria for which peers they interact with and operate only on their assigned subset.
+- The `PeerManager` controls connection policies such as reconnection for persistent peers, filtering, and connection limits.
+
+## Detailed Design
+
+### Reactor-Specific Peer Requirements
+
+This section defines which reactors require which types of peers and the behavior for existing reactors:
+
+#### Reactor Peer Requirements:
+1. **Consensus Reactor**:
+   - **Peer Types**: Validator nodes, full nodes participating in consensus
+   - **Criteria**: Nodes with validator keys or nodes that relay consensus messages
+   - **Connection Style**: persistent (for validator connections)
+
+2. **State Sync Reactor**:
+   - **Peer Types**: Full nodes with complete state, archive nodes
+   - **Criteria**: Nodes advertising state sync capability
+   - **Connection Style**: ad hoc / during sync
+   - **Network Requirements**: Requires larger amounts of peers to saturate network connection for faster state synchronization
+
+3. **Block Sync Reactor**:
+   - **Peer Types**: Full nodes, archive nodes with historical blocks
+   - **Criteria**: Nodes with block history and fast sync capability
+   - **Connection Style**: ad hoc / during sync
+   - **Network Requirements**: Requires larger amounts of peers to saturate network connection for efficient block downloading
+
+4. **Mempool Reactor**:
+   - **Peer Types**: All node types that accept transactions
+   - **Criteria**: None
+   - **Connection Style**: ad hoc
+
+5. **Evidence Reactor**:
+   - **Peer Types**: Validator nodes, full nodes
+   - **Criteria**: Nodes participating in consensus
+   - **Connection Style**: ad hoc
+
+#### Backward Compatibility for Existing Reactors:
+- **Default Behavior**: Reactors that don't specify peer criteria will receive all available peers (current behavior).
+- **Opt-in Migration**: Existing reactors can gradually adopt peer-specific criteria without breaking changes.
+- **Interface Compatibility**: Current reactor interfaces (`AddPeer`, `RemovePeer`) remain unchanged.
+
+### Workflow
+
+1. **Peer Addition**:
+    - The transport layer notifies the `PeerManager` when a new peer is connected.
+    - The `PeerManager` validates, filters, and, if approved, assigns the peer to one or multiple reactors based on their criteria.
+
+2. **Reactor-Specific Peer Interaction**:
+    - Each reactor interacts only with its assigned subset of peers.
+    - Reactors may independently add or remove specific peers from their subset.
+
+3. **Broadcasting**:
+    - When broadcasting messages, the `Switch` ensures only peers within the corresponding reactor's subset are targeted.
+
+4. **Reconnection and Removal**:
+    - The `PeerManager` implements policies for reconnection, persistent peers, and termination.
+    - If a peer is removed, all reactors interacting with that peer are notified.
```

Comment:
So in practice will there still be a global view somewhere of which reactor has which peer? Or do we just notify all reactors and the notification will be ignored if the reactor has no business with the removed peer?